api/types/authentication_authpreference_test.go

```diff
@@ -64,6 +64,81 @@ U9psmyPzK+Vsgw2jeRQ5JlKDyqE0hebfC1tvFu0CCrJFcw==
 -----END CERTIFICATE-----`
 )
 
+func TestU2FFacetValidation(t *testing.T) {
+	tests := []struct {
+		name    string
+		u2f     *types.U2F
+		wantErr bool
+	}{
+		{
+			name: "same-domain",
+			u2f: &types.U2F{
+				AppID: "https://foo.com",
+				Facets: []string{
+					"https://foo.com",
+					"https://foo.com:3080",
+				},
+			},
+			wantErr: false,
+		},
+		{
+			name: "same-domain-www-prefix",
+			u2f: &types.U2F{
+				AppID: "https://www.foo.com",
+				Facets: []string{
+					"https://foo.com",
+					"https://foo.com:3080",
+				},
+			},
+			wantErr: false,
+		},
+		{
+			name: "diff-origin",
+			u2f: &types.U2F{
+				AppID: "https://bar.com",
+				Facets: []string{
+					"https://foo.com",
+					"https://foo.com:3080",
+				},
+			},
+			wantErr: true,
+		},
+		{
+			name: "naked-protocol",
+			u2f: &types.U2F{
+				AppID: "https://foo.com",
+				Facets: []string{
+					"foo.com",
+					"foo.com:3080",
+				},
+			},
+			wantErr: true,
+		},
+		{
+			name: "localhost-naked-protocol",
+			u2f: &types.U2F{
+				AppID: "https://localhost",
+				Facets: []string{
+					"localhost",
+					"localhost:3080",
+				},
+			},
+			wantErr: false,
+		},
+	}
+
+	for _, test := range tests {
+		t.Run(test.name, func(t *testing.T) {
+			err := test.u2f.Check()
+			if test.wantErr {
+				require.Error(t, err)
+			} else {
+				require.NoError(t, err)
+			}
```

Comment:
nit: If you change the `wantErr bool` to `assertFn require.ErrorAssertionFunc` this can be simplified.

Reply:
This specific section yes, but it will introduce additional complexity in each test case. It seems simplest to keep like is.

Reply:
I'm with @jakule  on this one, I'd rather see `assertFn: require.NoError` in a test case and be able to remove the if here

Reply:
@jakule @rosstimothy Swapped to assertfn now